src/inc/til/rle.h

```diff
@@ -642,6 +654,13 @@ namespace til // Terminal Implementation Library. Also: "Today I Learned"
                 return { it, run_pos };
             }
 
+            void boost(It new_it, size_type new_total)
+            {
+                it = new_it;
+                total = new_total;
+                run_pos = 0;
+            }

```

Comment:
We could also change the scanner algorithm to be able to search both forwards and backward, like this code: https://github.com/microsoft/terminal/blob/aa256ad5c9656c8aac75b1cf40c0ecae869dd4c0/src/buffer/out/Row.cpp#L91-L126

If we then store the last iterator position and the cumulative offset of that iterator, we can construct the scanner immediately with that information. We can then also revert your other changes below and make it quite a bit simpler hopefully (apart from the replace-all fast-path).

I'm not sure if I explained my idea well, so please let me know if I should explain my idea in more detail!